### PR TITLE
Granite speech - minor fixes to support training with the HF trainer

### PR DESCRIPTION
**Pull request by avihu111**
# What does this PR do?

Minor updates to `granite_speech` to enable finetuning it with HF trainers.
- avoids a crash when trainers pass `padding=True` to the processor
- ensure all trainable parameters have gradients (bugfix) - remove `.data` from a forward call
- rename `melspec` to `mel_filters` to leverage [this](https://github.com/huggingface/transformers/blob/d5d007a1a0f0c11a726a54c8f00bd71825f84d02/src/transformers/feature_extraction_utils.py#L595), which avoids a crash on `save_pretrained`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

CC: @ArthurZucker @eustlb @alex-jw-brooks @avishaiElmakies @gsaon

### CODE REVIEW

File: src/transformers/models/granite_speech/feature_extraction_granite_speech.py [hunk 1]
```diff
@@ -50,15 +50,16 @@ def __init__(
         **kwargs,
     ):
         super().__init__(**kwargs)
+        self.sampling_rate = sampling_rate
```
Comment (alex-jw-brooks) [inline src/transformers/models/granite_speech/feature_extraction_granite_speech.py:53]: I think this isn't used currently

Comment (avihu111) [reply src/transformers/models/granite_speech/feature_extraction_granite_speech.py:53]: Right, it's not used. I added it to stay consistent with other audio feature extractors that have this property.

File: src/transformers/models/granite_speech/processing_granite_speech.py [hunk 2]
```diff
@@ -88,7 +88,9 @@ def __call__(
         else:
             audio_inputs = {}
 
-        text_inputs = self.tokenizer(prompt_strings, padding=True, **kwargs)
+        if "padding" not in kwargs:
```
Comment (avihu111) [inline src/transformers/models/granite_speech/processing_granite_speech.py:91]: avoids a crash when trainers pass `padding=True` to the processor

File: src/transformers/models/granite_speech/modeling_granite_speech.py [hunk 3]
```diff
@@ -92,7 +92,7 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         hidden_states = hidden_states.view(batch_size * nblocks, self.window_size, dim)
 
         query_output = self.qformer(
-            query_embeds=self.query.data,
+            query_embeds=self.query,
```
Comment (avihu111) [inline src/transformers/models/granite_speech/modeling_granite_speech.py:95]: Bugfix. When using `.data` this trainable parameter did not receive gradients.

File: src/transformers/models/granite_speech/feature_extraction_granite_speech.py [hunk 4]
```diff
@@ -50,15 +50,16 @@ def __init__(
...
             "sample_rate": sampling_rate,
             "n_fft": n_fft,
             "win_length": win_length,
             "hop_length": hop_length,
             "n_mels": n_mels,
         }
-        # Currently lazily initialized
-        self.melspec = None
+        requires_backends(self, ["torchaudio"])
+        self.mel_filters = torchaudio.transforms.MelSpectrogram(**self.melspec_kwargs)
```
Comment (avihu111) [inline src/transformers/models/granite_speech/feature_extraction_granite_speech.py:62]: removed the lazy init, and renamed it to `mel_filters`. This specific name avoids a crash when serializing the processor.

Comment (avihu111) [reply src/transformers/models/granite_speech/feature_extraction_granite_speech.py:62]: [see here](https://github.com/huggingface/transformers/blob/d5d007a1a0f0c11a726a54c8f00bd71825f84d02/src/transformers/feature_extraction_utils.py#L595)

**Review by avihu111 — COMMENTED** [review]
Added some comments on each change, giving relevant context

**Review by ArthurZucker — APPROVED** [review]
lgtm sorry for the delay!

**Comment by avihu111** [comment]
Awesome! Thanks @ArthurZucker!
